data/modules/MissionUtils/ShipBuilder.lua

```diff
@@ -0,0 +1,855 @@
+-- minThreat and randomChance rule parameters in each individual equipment rule.
+--
+-- See data/modules/MissionUtils/OutfitRules.lua for information on the fields
+-- applicable to an equipment rule.
+--
+-- Parameters:
+--
+--  template   - MissionUtils.ShipTemplate, the template containing equipment
+--               rules to evaluate.
+--
+--  shipConfig - HullConfig, the hull configuration to make a plan for.
+--
+--  threat     - number, controls the intended difficulty of the encounter. The
+--               threat value is used to limit which equipment can be installed
+--               and whether an equipment rule can be evaluated at all via the
+--               minThreat property.
+--
+-- Returns:
+--
+--  plan - table, an opaque structure containing information about the ship
+--         to spawn and the equipment to install as a result of evaluating the
+--         input template.
+--
+---@param template MissionUtils.ShipTemplate
+---@param shipConfig HullConfig
+---@param threat number
+---@return MissionUtils.ShipBuilder.ShipPlan
+function ShipBuilder.MakePlan(template, shipConfig, threat)
+
+	local hullThreat = ShipBuilder.GetHullThreat(shipConfig.id).total
+
+	local randomMod = template.randomModifier or 1.0
+
+	local shipPlan = ShipPlan:clone {
+		threat = hullThreat,
+		freeThreat = threat - hullThreat,
+		maxThreat = threat
+	}
+
+	shipPlan:SetConfig(shipConfig)
+
+	for _, rule in ipairs(template.rules) do
+
+		local canApplyRule = true
+
+		if rule.minThreat then canApplyRule = canApplyRule and threat >= rule.minThreat end
+		if rule.randomChance then canApplyRule = canApplyRule and Engine.rand:Number() < rule.randomChance * randomMod end
+
+		if canApplyRule then
+
+			if rule.slot then
+				ShipBuilder.ApplyEquipmentRule(shipPlan, rule, Engine.rand, hullThreat)
+			else
+				local equip = Equipment.Get(rule.equip)
+				assert(equip)
+
+				local equipThreat = ShipBuilder.ComputeEquipThreatFactor(equip, hullThreat)
+
+				if shipPlan.freeVolume >= equip.volume and shipPlan.freeThreat >= equipThreat then
+					shipPlan:AddEquipToPlan(equip)
+				end
+			end
+
+		end
+
+	end
+
+	shipPlan.label = template.label or Ship.MakeRandomLabel()
+
+	return shipPlan
+
+end
+
+-- Function: ApplyPlan
+--
+-- Apply the previously created plan to a concrete ship object of the
+-- previously-specified hull configuration, installing all equipment instances
+-- created by applying the equipment rules contained in the ship template.
+--
+-- This function ensures that equipment is properly installed in the order it
+-- was selected and items are properly distributed to sub-slots provided by
+-- installed equipment items.
+--
+-- Parameters:
+--
+--  ship     - Ship, the ship to install equipment into.
+--
+--  shipPlan - table, an opaque ship plan returned from a previous call to
+--             ShipBuilder.MakePlan.
+--
+---@param ship Ship
+---@param shipPlan MissionUtils.ShipBuilder.ShipPlan
+function ShipBuilder.ApplyPlan(ship, shipPlan)
+
```

Comment:
Might be an idea to add a sanity-check that the plan's hull config matches the ship hull.